### PR TITLE
Making the CLI package name be only major.minor version and updating the version of the runtime packages that we depend on.

### PR DESCRIPTION
**Pull request by livarcocc**
@MattGertz marked it for shiproom approval already because this change is a reaction to an upstream change in the runtime and asp.net. We need to take it.

**Issues fixed**

Fixes https://github.com/dotnet/cli/issues/9189

**Description of Issue**

Change the CLI's linux native installers to depend on version of the runtime and asp.net that are versioned and major.minor. Also update the CLI package names to be major.minor.

**Customer Impact**

Better package management by the native linux package managers.

**Risk**

Medium - this is changing the linux native installers of the CLI.

**Testing**

CLI CI.
Manually verified the dependencies of the new packages.

### CODE REVIEW

**Comment by livarcocc** [comment]
@dotnet-bot Test Debian8.2 x64 Debug Build

**Comment by JunTaoLuo** [comment]
Also note the filenames are different for upgradeable installers. For example:
Non-Upgradeable: dotnet-runtime-2.1.0-rtm-26507-03-x64.deb
Upgradeable: dotnet-runtime-2.1.0-rtm-26507-03-x64-upgrade.deb

I think this is going to require changes here: https://github.com/dotnet/cli/blob/master/build/BundledRuntimes.props#L10-L39.

**Comment by MattGertz** [comment]
Totally agree, thanks.

File: build/package/Installer.DEB.proj [hunk 1]
```diff
@@ -95,11 +95,11 @@
 
       <!-- Remove Shared Framework and Debian Packages -->
       <Exec Command="sudo dpkg -r $(SdkDebianPackageName)" />
-      <Exec Command="sudo dpkg -r $(AspNetCoreSharedFxDebianPackageName)" />
-      <Exec Command="sudo dpkg -r $(SharedFxDebianPackageName)" />
+      <Exec Command="sudo dpkg -r $(AspNetCoreSharedFxDebianPackageFileName)" />
```
Comment (JunTaoLuo) [inline build/package/Installer.DEB.proj]: We should double check that this works with the new package names. The package installed is aspnetcore-runtime-2.1 but I think this property is set to aspnetcore-runtime-2.1.0-rc1-30661 which is the non-upgradeable installer package name.

File: build/package/Installer.RPM.targets [hunk 2]
```diff
@@ -68,13 +68,13 @@
...
       <SDKTokenValue Include="%SHARED_HOST_RPM_VERSION%">
-        <ReplacementString>$(MicrosoftNETCoreAppPackageVersion)</ReplacementString>
+        <ReplacementString>$(MicrosoftNETCoreAppMajorMinorVersion)</ReplacementString>
       </SDKTokenValue>
       <SDKTokenValue Include="%ASPNETCORE_SHAREDFX_RPM_PACKAGE_NAME%">
         <ReplacementString>$(AspNetCoreSharedFxRpmPackageName)</ReplacementString>
       </SDKTokenValue>
       <SDKTokenValue Include="%ASPNETCORE_SHAREDFX_RPM_PACKAGE_VERSION%">
-        <ReplacementString>$(AspNetCoreVersion)</ReplacementString>
+        <ReplacementString>$(AspNetCoreMajorMinorVersion)</ReplacementString>
```
Comment (JunTaoLuo) [inline build/package/Installer.RPM.targets]: This is wrong. The package name is aspnetcore-runtime-2.1 but the version is 2.1.0 for this release. Future patches will be 2.1.1, etc.

**Comment by JunTaoLuo** [comment]
A couple of pitfalls to highlight:
Upgradeable deb installers have a tilde in their version number: e.g. dotnet-runtime-2.1 version: 2.1.0~rtm-26507 revision: 1
Upgradeable rpm installers always uses stable versions, but the build number is contained in the revision field: e.g. dotnet-runtime-2.1 version: 2.1.0 revision: 0.1.rtm-26507
The version used in installer package name is major.minor but the actual version of the installer is major.minor.patch and the version used in the filename is major.minor.patch-tag-buildnumber.
Also note the differences in version schemes between non-stabilized and stabilized versions. For example the revision field of stabilized rpm installers will be 1 instead of 0.1.tag-buildnumber.

**Comment by livarcocc** [comment]
@JunTaoLuo We use that format because of the additional shared framework, which is optional. It need to be in that format.

As for upgradable/non-upgradable packages, we use those packages for testing purposes and during our tests we are installing the packages directly and not doing upgrades. So, I believe it is ok to continue doing what we are doing.

**Comment by JunTaoLuo** [comment]
@livarcocc I think I'm misunderstanding something here. What do you mean by the format must stay the same because of the additional shared framework? Why not change the debian.config to something like:
```
    "debian_dependencies":{
        "%SHARED_FRAMEWORK_DEBIAN_PACKAGE_NAME%" : { "package_version": %SHARED_FRAMEWORK_DEBIAN_PACKAGE_VERSIONS% }%SHARED_FRAMEWORK_DEBIAN_PACKAGE_ADDITIONAL_DEPENDENCY%,
        "%ASPNETCORE_SHAREDFX_DEBIAN_PACKAGE_NAME%": { "package_version": %ASPNETCORE_SHAREDFX_DEBIAN_PACKAGE_VERSION% }
    }
```
?

Also what I mean is that the tests should be updated to try installing the upgradeable installers. You don't necessarily have to test upgrade behaviours (though that would be nice) but if you don't change the name of the installer files you download, you will still be testing with non-upgradeable installers.

**Comment by livarcocc** [comment]
The package_version in the debian_config.json file is optional. I imagine in this case, we just need to declare a dependency on a 2.1 package, for instance and it should all just work. I would prefer not to change it at this point or conflate this change (upgradable packages) with that.

Are you saying that the package_version information is required for upgradable packages? If so, how is that so?

As for the upgradable packages, what is the different between -upgradable and the regular packages? So that I can see if it is worth to change it along with this change as well.

**Comment by livarcocc** [comment]
@JunTaoLuo I see that only the -upgradable packages have the new version format. You are right. We need to update this part.

**Comment by wli3** [comment]
help me understand. the reason packages are not "upgradable" because in manifest of deb/rpm CLI.deb depend on a package name with version in it, say `aspnetcore-runtime-2.1`. And to make it upgradable, we need to depend on package name "aspnetcore-runtime" and version should be >=2.1. It that correct?

**Comment by wli3** [comment]
at the same name, CLI need to change its package name from dotnet-sdk-2.1 to 
package name: dotnet-sdk 
version: 2.1

**Comment by wli3** [comment]
help discussion. The result of dpkg-deb -I 
```
dpkg-deb -I aspnetcore-runtime-2.2.0-preview1-34148-x64-upgrade.deb
 new debian package, version 2.0.
 size 22121078 bytes: control archive=7637 bytes.
     703 bytes,    11 lines      control
   33817 bytes,   235 lines      md5sums
 Package: aspnetcore-runtime-2.2
 Version: 2.2.0~preview1-34148-1
 Architecture: amd64
 Maintainer: Microsoft <nugetaspnet@microsoft.com>
 Installed-Size: 72541
 Depends: libc6 (>= 2.14), dotnet-runtime-2.1 (>= 2.1.0~rc1)
 Section: devel
 Priority: standard
 Homepage: https://www.asp.net/
 Description: Microsoft ASP.NET Core 2.2.0-preview1-34148 Shared Framework
  Shared Framework for hosting of Microsoft ASP.NET Core applications. It is open source, cross-platform and is supported by Microsoft. We hope you enjoy using it! If you do, please consider joining the active community of developers that are contributing to the project on GitHub (https://github.com/aspnet/home). We happily accept issues and PRs.
```

**Comment by wli3** [comment]
I need to know more about how we define upgradable. Looks like that aspnetcore-runtime cannot work with next minor version of runtime according to deb manifest. However, according to the runtime policy. It should. Also should we say `dotnet-runtime (>= 2.1.0~rc1), dotnet-runtime (<< 3.0.0)`?

https://askubuntu.com/questions/766846/version-range-in-debian-control

**Comment by wli3** [comment]
@JunTaoLuo thank you. I think I am on the same page now.

**Comment by wli3** [comment]
Agree something more need to be done on debian.config

current dpkg-deb -I don't have the version range. We need to add something

` Depends: libc6 (>= 2.4), libgcc1 (>= 1:4.1.1), libstdc++6 (>= 4.8), dotnet-runtime-2.1.0-preview3-26411-06 !!! no version range here !!!, aspnetcore-runtime-2.1.0-preview2-30475`

Full dpkg-deb -I:
```
dpkg-deb -I  dotnet-sdk-latest-x64.deb
 new debian package, version 2.0.
 size 83943916 bytes: control archive=43619 bytes.
     849 bytes,    11 lines      control
  179802 bytes,  1327 lines      md5sums
    1451 bytes,    27 lines   *  postinst             #!/usr/bin/env
 Package: dotnet-sdk-2.1.400-preview-008659
 Version: 2.1.400-preview-008659-1
 Architecture: amd64
 Maintainer: Microsoft <dotnetcore@microsoft.com>
 Installed-Size: 220295
 Depends: libc6 (>= 2.4), libgcc1 (>= 1:4.1.1), libstdc++6 (>= 4.8), dotnet-runtime-2.1.0-preview3-26411-06, aspnetcore-runtime-2.1.0-preview2-30475
 Section: devel
 Priority: standard
 Homepage: https://dotnet.github.io/core
 Description: Microsoft .NET Core SDK 2.1.400 - preview
  .NET Core is a development platform that you can use to build command-line applications, microservices and modern websites. It is open source, cross-platform and is supported by Microsoft. We hope you enjoy using it! If you do, please consider joining the active community of developers that are contributing to the project on GitHub (https://github.com/dotnet/core). We happily accept issues and PRs.
```

**Comment by JunTaoLuo** [comment]
@livarcocc I would recommend specifying a version range for the dotnet and aspnetcore runtime dependencies. This would keep the sdk and the runtime it carries in sync. Suppose the user is upgrading from 2.1.0 to 2.1.1 of  the cli package via `apt-get install dotnet-cli-2.1`. If no version range is specified, the installation will succeed and only upgrade the cli whereas if you specified the version range, the apt-get will also upgrade dotnet-runtime and other dependencies to 2.1.1 (or fail and tell you that dependencies also need to be upgraded depending on the options you specify). I think we would want to avoid the case where the user has the 2.1.0 runtime but the 2.1.1 sdk. 

Note that in the case that the user runs `apt-get upgrade` everything should work fine even without the versioned dependencies since all packages including the cli's dependencies are upgraded.

**Comment by leecow** [comment]
@JunTaoLuo , your description is accurate. @wli3 , you do need to provide some range parameter to the runtime dependency in order for this to work. Here's what we have in the runtime package, as an example. 

```
Depends: dotnet-runtime-deps-2.1 (>= 2.1.0~rtm-26508-02), dotnet-hostfxr-2.1 (>= 2.1.0~rtm-26508-02)
```

**Comment by livarcocc** [comment]
@leecow I am the one working on the PR. I will work on adding the ranges in.

**Comment by livarcocc** [comment]
Added the version range:

```
dpkg-deb -I bin/2/linux-x64/packages/dotnet-sdk-2.1.300-rtm-008826-x64.deb 
 new debian package, version 2.0.
 size 87751096 bytes: control archive=37647 bytes.
     830 bytes,    11 lines      control              
  173766 bytes,  1322 lines      md5sums              
    1447 bytes,    27 lines   *  postinst             #!/usr/bin/env
 Package: dotnet-sdk-2.1
 Version: 2.1.300-rtm-008826-1
 Architecture: amd64
 Maintainer: Microsoft <dotnetcore@microsoft.com>
 Installed-Size: 222947
 Depends: libc6 (>= 2.4), libgcc1 (>= 1:4.1.1), libstdc++6 (>= 4.8), dotnet-runtime-2.1 (>= 2.1.0-rtm-26508-02), aspnetcore-runtime-2.1 (>= 2.1.0-rtm-30752)
 Section: devel
 Priority: standard
 Homepage: https://dotnet.github.io/core
 Description: Microsoft .NET Core SDK 2.1.300 - rtm
  .NET Core is a development platform that you can use to build command-line applications, microservices and modern websites. It is open source, cross-platform and is supported by Microsoft. We hope you enjoy using it! If you do, please consider joining the active community of developers that are contributing to the project on GitHub (https://github.com/dotnet/core). We happily accept issues and PRs.
```

**Comment by livarcocc** [comment]
@dotnet-bot Test Ubuntu x64 Release Build

**Comment by wli3** [comment]
deb looks good. Do we have the same requirement for RPM?

**Comment by wli3** [comment]
make sure my understanding. RPM has this version range all the time. So we only need to change the package name

**Comment by JunTaoLuo** [comment]
@livarcocc, I what are the file names of the cli deb and rpm installers? I think you should keep the major.minor.patch version inside the filename like we did for the other installers so it's easy to tell which version of the cli it contains. i.e. dotnet-sdk-2.1.0.rpm installs dotnet-sdk-2.1 with version of 2.1.0.

**Pull request merged**